docs/configuration.md
# Configuration

## Account Section

Configure your account settings in the `account.json` file under the `account` section. This includes API keys, account names, and other related settings.

### Parameters:

- **name**: Exchange name
- **api_key**: Your exchange API key.
- **api_secret**: Your exchange API secret.
- **account_name**: A name to identify your account configuration.

## Accounts

Configure your accounts using 'account.json' in the configs folder.

```json
{
    "exchanges": [
        {
            "name": "bybit",
            "account_name": "account_1",
            "api_key": "your_api_key",
            "api_secret": "your_api_secret"
        },
        {
            "name": "bybit_spot",
            "account_name": "account_2",
            "api_key": "your_api_key",
            "api_secret": "your_api_secret"
        },
        {
            "name": "bybit_unified",
            "account_name": "account_3",
            "api_key": "your_api_key",
            "api_secret": "your_api_secret"
        },
        {
            "name": "bitget",
            "account_name": "account_4",
            "api_key": "your_api_key",
            "api_secret": "your_api_secret"     
        }
    ]
}
```

## Bot Section

Configure your bot settings in the `config.json` file under the `bot` section. This includes strategy parameters, risk management settings, and other related settings.

### Risk management parameters

- **Exchange leverage is MAX CROSS on all pairs by default**: Changing user_defined_leverage does not change leverage on exchange instead it changes the amount the bot is allowed to use.
- **user_defined_leverage_long**: This parameter defines the leverage to use for long positions. Leverage is a multiplier that increases the size of the position and the potential profit/loss. For example, a leverage of 10 means that for every $1 you have, you can place a trade worth $10. Be careful with this setting as higher leverage also increases the risk.
- **user_defined_leverage_short**: This parameter defines the leverage to use for short positions. Similar to `user_defined_leverage_long`, it's a multiplier that increases the size of the position and the potential profit/loss.

### Risk Management Explanation

Risk Management plays a crucial role in trading strategies, as it helps protect the trader's capital and limits potential losses. In the provided code snippet, there are several parameters and calculations related to risk management. Let's dive into each of them and provide examples to clarify their usage.

- **user_defined_leverage_long** and **user_defined_leverage_short**:
  - These parameters allow the user to specify the leverage for long and short positions, respectively.
  - Leverage is a multiplier that increases the size of the position and the potential profit/loss.
  - Example:
    - If `user_defined_leverage_long` is set to 5, it means that for every $1 of the trader's capital, they can open a long position worth $5.
    - Similarly, if `user_defined_leverage_short` is set to 3, it means that for every $1 of the trader's capital, they can open a short position worth $3.

- **wallet_exposure_limit_long** and **wallet_exposure_limit_short**:
  - These parameters define the maximum percentage of the trader's total equity that can be allocated to long and short positions, respectively.
  - They help limit the overall exposure to risk by controlling the amount of capital deployed in each position.
  - Example:
    - If `wallet_exposure_limit_long` is set to 0.1 (10%), it means that the maximum amount allocated to long positions cannot exceed 10% of the total equity.
    - Similarly, if `wallet_exposure_limit_short` is set to 0.05 (5%), it means that the maximum amount allocated to short positions cannot exceed 5% of the total equity.

- **Effective Leverage Calculation**:
  - The effective leverage used in trading is calculated as `total_equity * wallet_exposure_limit * leverage_used`.
  - This calculation determines the maximum position value based on the trader's total equity, wallet exposure limit, and the selected leverage.
  - Example:
    - Let's say a trader has a total equity of $10,000, `wallet_exposure_limit_long` is set to 0.1 (10%), and `user_defined_leverage_long` is set to 5.
    - The maximum position value for long positions would be: $10,000 * 0.1 * 5 = $5,000.
    - This means that the trader can open long positions with a total value of up to $5,000.

- **enforce_full_grid**:
  - This parameter determines whether the trading strategy should enforce a full grid of positions or allow partial grid usage.
  - If `enforce_full_grid` is set to `True`, the code calculates the required notional amount by dividing the maximum position value by the number of levels.
  - If `enforce_full_grid` is set to `False`, the code uses the maximum position value as the required notional amount.
  - Example:
    - Let's say the maximum position value is $5,000, and the trading strategy uses 5 levels.
    - If `enforce_full_grid` is `True`, the required notional amount for each level would be: $5,000 / 5 = $1,000.
    - If `enforce_full_grid` is `False`, the required notional amount would be the full $5,000.

- **Adjusting Maximum Position Value**:
  - The code takes into account the current position value (long or short) and adjusts the maximum position value accordingly.
  - It calculates the current position value by multiplying the position quantity (long or short) with the respective best ask or bid price.
  - The adjusted maximum position value is then calculated by subtracting the current position value from the original maximum position value.
  - Example:
    - Let's say the maximum position value is $5,000, and the trader currently holds a long position with a quantity of 2 units at a best ask price of $100.
    - The current position value would be: 2 * $100 = $200.
    - The adjusted maximum position value would be: $5,000 - $200 = $4,800.
    - This means that the trader can open additional positions with a total value of up to $4,800.

These risk management parameters and calculations help traders control their exposure to risk, limit potential losses, and manage their positions effectively. By adjusting these parameters based on their risk tolerance and trading strategy, traders can optimize their risk management approach and protect their capital.

### Linear Grid Strategy Parameters

Here are the parameters for the `linear_grid` strategy in your `config.json` file:

- **max_qty_percent_long**: Defines the maximum percentage of your wallet balance that can be used for long positions.
- **max_qty_percent_short**: Defines the maximum percentage of your wallet balance that can be used for short positions.
- **auto_reduce_cooldown_enabled**: Boolean parameter that determines whether the auto-reduce cooldown feature is enabled. This feature automatically reduces the size of your positions after a certain period.
- **auto_reduce_cooldown_start_pct**: Defines the percentage at which the auto-reduce cooldown feature starts.
- **wallet_exposure_limit_long**: Defines the maximum exposure limit for long positions as a percentage of your wallet balance.
- **wallet_exposure_limit_short**: Defines the maximum exposure limit for short positions as a percentage of your wallet balance.
- **levels**: Defines the number of grid levels, representing the total number of buy and sell orders placed above and below the current price level.
- **strength**: Defines the strength of the grid. A higher value means the grid levels are spaced further apart.
- **outer_price_distance**: Defines the distance from the current price to the outermost grid levels.
- **min_outer_price_distance**: Defines the minimum distance from the current price to the outermost grid levels.
- **max_outer_price_distance**: Defines the maximum distance from the current price to the outermost grid levels.
- **long_mode**: Boolean parameter that determines whether the bot can open long positions.
- **short_mode**: Boolean parameter that determines whether the bot can open short positions.
- **reissue_threshold**: Defines the threshold at which the bot will reissue orders that have been partially filled.
- **buffer_percentage**: Defines the buffer percentage for the grid. This is the percentage of the price range kept empty between the outermost grid levels and the upper and lower price limits.
- **initial_entry_buffer_pct**: Defines the buffer percentage for the initial entry.
- **min_buffer_percentage**: Defines the minimum buffer percentage for the grid.
- **max_buffer_percentage**: Defines the maximum buffer percentage for the grid.
- **enforce_full_grid**: Boolean parameter that determines whether the bot should always maintain a full grid of orders.
- **min_buffer_percentage_ar**: Defines the minimum buffer percentage for auto-reduce.
- **max_buffer_percentage_ar**: Defines the maximum buffer percentage for auto-reduce.
- **upnl_auto_reduce_threshold_long**: Defines the unrealized profit and loss (UPNL) threshold for auto-reducing long positions.
- **upnl_auto_reduce_threshold_short**: Defines the UPNL threshold for auto-reducing short positions.
- **failsafe_enabled**: Boolean parameter that determines whether the failsafe feature is enabled. This feature automatically closes all positions if the UPNL reaches a certain threshold.
- **failsafe_start_pct**: Defines the percentage at which the failsafe feature starts.
- **long_failsafe_upnl_pct**: Defines the UPNL percentage for the long failsafe.
- **short_failsafe_upnl_pct**: Defines the UPNL percentage for the short failsafe.

### Example Configuration Snippet

```json
{
  "api": {
      "filename": "quantdatav2_bybit.json",
      "mode": "remote",
      "url": "https://api.quantumvoid.org/volumedata/",
      "data_source_exchange": "bybit"
  },
  "bot": {
      "bot_name": "your_bot_name",
      "symbol": "BTCUSDT",
      "volume_check": false,
      "min_distance": 0.15,
      "min_volume": 10000,
      "wallet_exposure_limit": 0.001,
      "user_defined_leverage_long": 8,
      "user_defined_leverage_short": 5,
      "upnl_profit_pct": 0.0022,
      "max_upnl_profit_pct": 0.0029,
      "auto_reduce_enabled": false,
      "auto_reduce_start_pct": 0.068,
      "entry_during_autoreduce": false,
      "stoploss_enabled": false,
      "stoploss_upnl_pct": 0.05,
      "liq_stoploss_enabled": false,
      "liq_price_stop_pct": 0.50,
      "percentile_auto_reduce_enabled": false,
      "upnl_threshold_pct": 0.50,
      "max_pos_balance_pct": 0.50,
      "auto_reduce_wallet_exposure_pct": 0.20,
      "auto_reduce_maxloss_pct": 0.30,
      "auto_reduce_marginbased_enabled": false,
      "hedge_ratio": 0.10,
      "hedge_price_difference_threshold": 0.10,
      "auto_leverage_upscale": false,
      "test_orders_enabled": false,
      "max_usd_value": 50,
      "min_qty_threshold": 0,
      "long_liq_pct": 0.05,
      "short_liq_pct": 0.05,
      "MaxAbsFundingRate": 0.0002,
      "blacklist": ["BTCUSDT", "ETHUSDT"],
      "whitelist": [],
      "dashboard_enabled": false,
      "shared_data_path": "data/",
      "linear_grid": {
          "max_qty_percent_long": 5,
          "max_qty_percent_short": 5,
          "auto_reduce_cooldown_enabled": false,
          "auto_reduce_cooldown_start_pct": 0.051,
          "wallet_exposure_limit_long": 0.001,
          "wallet_exposure_limit_short": 0.001,
          "levels": 3,
          "strength": 1.4,
          "outer_price_distance": 0.059,
          "min_outer_price_distance": 0.019,
          "max_outer_price_distance": 0.039,
          "long_mode": true,
          "short_mode": true,
          "reissue_threshold": 0.001,
          "buffer_percentage": 0.10,
          "initial_entry_buffer_pct": 0.0001,
          "min_buffer_percentage": 0.0035,
          "max_buffer_percentage": 0.010,
          "enforce_full_grid": true,
          "min_buffer_percentage_ar": 0.002,
          "max_buffer_percentage_ar": 0.004,
          "upnl_auto_reduce_threshold_long": 10.0,
          "upnl_auto_reduce_threshold_short": 10.0,
          "failsafe_enabled": false,
          "failsafe_start_pct": 0.05,
          "long_failsafe_upnl_pct": 10.0,
          "short_failsafe_upnl_pct": 10.0
      },
      "hotkeys": {
          "hotkeys_enabled": false,
          "enter_long": "1",
          "take_profit_long": "2",
          "enter_short": "3",
          "take_profit_short": "4"
      }
  },
  "exchanges": [
      {
          "name": "bybit",
          "account_name": "account_1",
          "symbols_allowed": 10 
      },
      {
          "name": "bybit_spot",
          "account_name": "account_2",
          "symbols_allowed": 5
      },
      {
          "name": "bybit_unified",
          "account_name": "account_3",
          "symbols_allowed": 5
      },
      {
          "name": "bitget",
          "account_name": "account_4",
          "symbols_allowed": 5
      } 
  ],
  "logger": {
      "level": "info"
  },
  "messengers": {
      "discord": {
          "active": false,
          "embedded_messages": true,
          "messenger_type": "discord",
          "webhook_url": "https://discord.com/api/webhooks/your_webhook_id/your_webhook_token"
      },
      "telegram": {
          "active": false,
          "embedded_messages": true,
          "messenger_type": "telegram",
          "bot_token": "your_bot_token",
          "chat_id": "your_chat_id"
      }
  }
}
```



config.py
from __future__ import annotations
from typing import List, Optional, Union, Dict

import os
from pathlib import Path
import json
from enum import Enum

from pydantic import BaseModel, HttpUrl, ValidationError, validator, DirectoryPath

from directionalscalper.core.strategies.logger import Logger
logging = Logger(logger_name="Configuration", filename="Configuration.log", stream=True)

VERSION = "v2.9.9"

class Exchanges(Enum):
    BYBIT = "bybit"

class Messengers(Enum):
    DISCORD = "discord"
    TELEGRAM = "telegram"

class API(BaseModel):
    filename: str = "quantdatav2.json"
    mode: str = "remote"
    url: str = "https://api.quantumvoid.org/volumedata/"
    data_source_exchange: str = "bybit"

class Hotkeys(BaseModel):
    hotkeys_enabled: bool = False
    enter_long: str = "1"
    take_profit_long: str = "2"
    enter_short: str = "3"
    take_profit_short: str = "4"

class Bot(BaseModel):
    bot_name: str
    volume_check: bool = True
    min_distance: float = 0.15
    min_volume: int = 15000
    wallet_exposure_limit: float = 0.0001
    user_defined_leverage_long: float = 0
    user_defined_leverage_short: float = 0
    upnl_profit_pct: float = 0.003
    max_upnl_profit_pct: float = 0.004
    stoploss_enabled: bool = False
    stoploss_upnl_pct: float = 0.070
    liq_stoploss_enabled: bool = False
    liq_price_stop_pct: float = 0.50
    percentile_auto_reduce_enabled: bool = False
    auto_reduce_enabled: bool = False
    auto_reduce_start_pct: float = 0.098
    upnl_threshold_pct: float = 0.10
    max_pos_balance_pct: float = 0.20
    auto_reduce_maxloss_pct: float = 0.50
    auto_reduce_marginbased_enabled: bool = False
    auto_reduce_wallet_exposure_pct: float = 0.10
    entry_during_autoreduce: bool = True
    hedge_ratio: float = 0.26
    hedge_price_difference_threshold: float = 0.15
    auto_leverage_upscale: bool = False
    min_qty_threshold: float = 0
    symbol: str
    long_liq_pct: float = 0.05
    short_liq_pct: float = 0.05
    MaxAbsFundingRate: float = 0.0002
    wallet_exposure: float = 1.00
    test_orders_enabled: bool = False
    max_usd_value: Optional[float] = None
    blacklist: List[str] = []
    whitelist: List[str] = []
    dashboard_enabled: bool = False
    shared_data_path: Optional[str] = None
    linear_grid: Optional[dict] = None
    hotkeys: Hotkeys

    @validator('hotkeys')
    def validate_hotkeys(cls, value):
        if not value:
            raise ValueError("hotkeys must be provided and valid")
        return value

    @validator('linear_grid')
    def validate_linear_grid(cls, value):
        if value is None:
            raise ValueError("linear_grid must be a dictionary - check example config")
        return value

    @validator("upnl_profit_pct")
    def minimum_upnl_profit_pct(cls, v):
        if v < 0.0:
            raise ValueError("upnl_profit_pct must be greater than 0")
        return v
    
    @validator("max_upnl_profit_pct")
    def minimum_max_upnl_profit_pct(cls, v):
        if v < 0.0:
            raise ValueError("max_upnl_profit_pct must be greater than 0")
        return v
    
    @validator("min_volume")
    def minimum_min_volume(cls, v):
        if v < 0.0:
            raise ValueError("min_volume must be greater than 0")
        return v

    @validator("min_distance")
    def minimum_min_distance(cls, v):
        if v < 0.0:
            raise ValueError("min_distance must be greater than 0")
        return v

    @validator("long_liq_pct")
    def minimum_long_liq_pct(cls, v):
        if v < 0.0:
            raise ValueError("long_liq_pct must be greater than 0")
        return v

    @validator("short_liq_pct")
    def minimum_short_liq_pct(cls, v):
        if v < 0.0:
            raise ValueError("short_liq_pct must be greater than 0")
        return v

    @validator('test_orders_enabled')
    def check_test_orders_enabled_is_bool(cls, v):
        if not isinstance(v, bool):
            raise ValueError("test_orders_enabled must be a boolean")
        return v
    
    @validator('auto_reduce_enabled')
    def check_auto_reduce_enabled_is_bool(cls, v):
        if not isinstance(v, bool):
            raise ValueError("auto_reduce_enabled must be a boolean")
        return v

    @validator('liq_stoploss_enabled')
    def check_liq_stoploss_enabled_is_bool(cls, v):
        if not isinstance(v, bool):
            raise ValueError("liq_stoploss_enabled must be a boolean")
        return v

    @validator('liq_price_stop_pct')
    def validate_liq_price_stop_pct(cls, v):
        if v < 0.0 or v > 1.0:
            raise ValueError("liq_price_stop_pct must be between 0.0 and 1.0")
        return v

    @validator('auto_reduce_start_pct')
    def validate_auto_reduce_start_pct(cls, v):
        if v < 0.0 or v > 1.0:
            raise ValueError("auto_reduce_start_pct must be between 0.0 and 1.0")
        return v
    
    @validator('upnl_threshold_pct')
    def validate_upnl_threshold_pct(cls, v):
        if v < 0.0 or v > 1.0:
            raise ValueError("upnl_threshold_pct must be between 0.0 and 1.0")
        return v
    
    @validator('auto_reduce_maxloss_pct')
    def validate_auto_reduce_maxloss_pct(cls, v):
        if v < 0.0 or v > 1.0:
            raise ValueError("auto_reduce_maxloss_pct must be between 0.0 and 1.0")
        return v

    @validator('entry_during_autoreduce')
    def check_entry_during_autoreduce_is_bool(cls, v):
        if not isinstance(v, bool):
            raise ValueError("entry_during_autoreduce must be a boolean")
        return v

    @validator('auto_reduce_marginbased_enabled')
    def check_auto_reduce_marginbased_enabled_is_bool(cls, v):
        if not isinstance(v, bool):
            raise ValueError("auto_reduce_marginbased_enabled must be a boolean")
        return v

    @validator('auto_reduce_wallet_exposure_pct')
    def validate_auto_reduce_wallet_exposure_pct(cls, v):
        if v < 0.0 or v > 1.0:
            raise ValueError("auto_reduce_wallet_exposure_pct must be between 0.0 and 1.0")
        return v

    @validator('percentile_auto_reduce_enabled')
    def check_percentile_auto_reduce_enabled_is_bool(cls, v):
        if not isinstance(v, bool):
            raise ValueError("percentile_auto_reduce_enabled must be a boolean")
        return v

    @validator('max_pos_balance_pct')
    def validate_max_pos_balance_pct(cls, v):
        if v < 0.0:
            raise ValueError("max_pos_balance_pct must be between 0.0 and 1.0")
        return v

    @validator('auto_leverage_upscale')
    def check_auto_leverage_upscale_is_bool(cls, v):
        if not isinstance(v, bool):
            raise ValueError("auto_leverage_upscale must be a boolean")
        return v

    @validator('volume_check')
    def check_volume_check_is_bool(cls, v):
        if not isinstance(v, bool):
            raise ValueError("volume_check must be a boolean")
        return v

class Exchange(BaseModel):
    name: str
    account_name: str
    api_key: str
    api_secret: str
    passphrase: Optional[str] = None
    symbols_allowed: int = 12

class Logger(BaseModel):
    level: str = "info"

    @validator("level")
    def check_level(cls, v):
        levels = ["notset", "debug", "info", "warn", "error", "critical"]
        if v not in levels:
            raise ValueError(f"Log level must be in {levels}")
        return v

class Discord(BaseModel):
    active: bool = False
    embedded_messages: bool = True
    messenger_type: str = Messengers.DISCORD.value  # type: ignore
    webhook_url: HttpUrl

    @validator("webhook_url")
    def minimum_divider(cls, v):
        if not str(v).startswith("https://discord.com/api/webhooks/"):
            raise ValueError(
                "Discord webhook begins: https://discord.com/api/webhooks/"
            )
        return v

class Telegram(BaseModel):
    active: bool = False
    embedded_messages: bool = True
    messenger_type: str = Messengers.TELEGRAM.value  # type: ignore
    bot_token: str
    chat_id: str

class Config(BaseModel):
    api: API
    bot: Bot
    exchanges: List[Exchange]
    logger: Logger
    messengers: Dict[str, Union[Discord, Telegram]]

def resolve_shared_data_path(relative_path: str) -> Path:
    # Get the directory of the config file
    base_path = Path(__file__).parent

    # Resolve the relative path
    absolute_path = (base_path / relative_path).resolve()

    if not absolute_path.exists():
        raise ValueError(f"Shared data path does not exist: {absolute_path}")

    return absolute_path

def load_json(file_path: Path) -> dict:
    with open(file_path, 'r') as file:
        return json.load(file)

def load_config(config_path: Path, account_path: Path) -> Config:
    config_data = load_json(config_path)
    account_data = load_json(account_path)

    # Merge account data into config data
    for exchange in config_data['exchanges']:
        for account in account_data['exchanges']:
            if exchange['name'] == account['name'] and exchange['account_name'] == account['account_name']:
                exchange.update({
                    'api_key': account['api_key'],
                    'api_secret': account['api_secret'],
                    'passphrase': account.get('passphrase', None)
                })
                break

    return Config(**config_data)

def get_exchange_name(cli_exchange_name):
    if cli_exchange_name:
        return cli_exchange_name
    else:
        with open('config.json') as file:
            data = json.load(file)
            return data['exchanges'][0]['name']

def get_exchange_credentials(exchange_name, account_name):
    with open('config.json') as file:
        data = json.load(file)
        exchange_data = None
        for exchange in data['exchanges']:
            if exchange['name'] == exchange_name and exchange['account_name'] == account_name:
                exchange_data = exchange
                break
        if exchange_data:
            api_key = exchange_data['api_key']
            secret_key = exchange_data['api_secret']
            passphrase = exchange_data.get('passphrase')
            symbols_allowed = exchange_data.get('symbols_allowed', 12)  # Default to 12 if not specified

            # Logging the symbols_allowed value
            if 'symbols_allowed' in exchange_data:
                logging.info(f"Retrieved symbols_allowed for {exchange_name}: {symbols_allowed}")
            else:
                logging.warning(f"symbols_allowed not found for {exchange_name}. Defaulting to 12.")
            
            return api_key, secret_key, passphrase, symbols_allowed
        else:
            raise ValueError(f"Account {account_name} for exchange {exchange_name} not found in the config file.")
